native/python/pyjp_buffer.cpp

```diff
@@ -0,0 +1,139 @@
+#include "jpype.h"
+#include "pyjp.h"
+#include "jp_buffer.h"
+#include "jp_buffertype.h"
+
+#ifdef __cplusplus
+extern "C"
+{
+#endif
+
+static void PyJPBuffer_dealloc(PyJPBuffer *self)
+{
+	JP_PY_TRY("PyJPBuffer_dealloc");
+	delete self->m_Buffer;
+	Py_TYPE(self)->tp_free(self);
+	JP_PY_CATCH();
+}
+
+static PyObject *PyJPBuffer_repr(PyJPBuffer *self)
+{
+	JP_PY_TRY("PyJPBuffer_repr");
+	JPContext *context = PyJPModule_getContext();
+	JPJavaFrame frame(context);
+	if (self->m_Buffer == NULL)
+		JP_RAISE(PyExc_ValueError, "Null array");
+	stringstream sout;
+
+	sout << "<java beffer " << self->m_Buffer->getClass()->toString() << ">";
+	return JPPyString::fromStringUTF8(sout.str()).keep();
+	JP_PY_CATCH(0);
+}
+
+static void PyJPBuffer_releaseBuffer(PyJPBuffer *self, Py_buffer *view)
+{
+	JP_PY_TRY("PyJPBufferPrimitive_releaseBuffer");
+	JP_PY_CATCH();
```

Comment:
I honestly have no clue what I am supposed to put here.  The object dereference is done automatically and we don't have to resource track.  Thus I left a stub for now.

Reply:
The buffer capi documentation says that the getBuffer and release functions are a pair like free/malloc. So if the view has been allocated via getBuffer it has to be cleared by calling release. Why is it de-referenced automatically in every case?

Reply:
This type is not references as unlike all of the other buffers we use this is direct buffers only so the memory can't move and never needs to be pinned.  The object counter is the PyJPBuffer and as far as I can tell this is automatically done.  So not much appears need to be here.

Reply:
But you need to register the function as part of a mandatory set of the Python buffer protocol? 
A comment about this in the code wouldn't hurt anybody.
Also I don't understand the try catch as nothing will be thrown in this block, right?

Reply:
Yeah as far as I can tell from looking at the C code it was a requirement.  There were places where it appeared not to be checking if the release was null and I wasn't going to risk it.   

The worthless try/catch in this block is for the tracer system not anything meaningful to the user.  In the case where I can't catch an error in the debugger and I have to find all the cases each object was born and died I have to use the tracer.  Unfortunately the JP_TRACE_OUT pattern implies that it will rethrow the exception which makes some C++ compilers unhappy because a extern "C" code can't have an exception.  Thus the next safest code to use the tracer is JP_PY_TRY/JP_PY_CATCH which should appear in all extern "C" Python calls.   I know that it ends up with some really odd code because it appears like we are handling exceptions in routines that don't have any C++ code or exceptions.   It also does one other function, which is these blocks can serve as failure points for instrumentation.   Yes, I release we have a lot of stuff hooked up to those macros (tracing, error handling, and instrumentation) but splitting them out just makes the code blotted and less readable.